src/Java.Interop/Java.Interop/JniRuntime.JniTypeManager.cs

```diff
@@ -32,48 +33,80 @@ protected virtual void Dispose (bool disposing)
 				disposed    = true;
 			}
 
+			[MethodImpl (MethodImplOptions.AggressiveInlining)]
 			void AssertValid ()
 			{
 				if (!disposed)
 					return;
 				throw new ObjectDisposedException (nameof (JniTypeManager));
 			}
 
+			// NOTE: This method needs to be kept in sync with GetTypeSignatures()
+			// This version of the method has removed IEnumerable for performance reasons.
 			public JniTypeSignature GetTypeSignature (Type type)
 			{
 				AssertValid ();
 
-				return GetTypeSignatures (type).FirstOrDefault ();
+				if (type == null)
+					throw new ArgumentNullException (nameof (type));
+				if (type.ContainsGenericParameters)
+					throw new ArgumentException ($"'{type}' contains a generic type definition. This is not supported.", nameof (type));
+
+				type = GetUnderlyingType (type, out int rank);
+
+				foreach (var mapping in JniBuiltinTypeNameMappings.Value) {
+					if (mapping.Key == type) {
+						var r = mapping.Value;
+						return r.AddArrayRank (rank);
+					}
+				}
+
+				foreach (var mapping in JniBuiltinArrayMappings.Value) {
+					if (mapping.Key == type) {
+						var r = mapping.Value;
+						return r.AddArrayRank (rank);
+					}
+				}
+
+				var name = type.GetCustomAttribute<JniTypeSignatureAttribute> (inherit: false);
+				if (name != null) {
+					return new JniTypeSignature (name.SimpleReference, name.ArrayRank + rank, name.IsKeyword);
+				}
+
+				var isGeneric = type.IsGenericType;
+				var genericDef = isGeneric ? type.GetGenericTypeDefinition () : type;
+				if (isGeneric) {
+					if (genericDef == typeof (JavaArray<>) || genericDef == typeof (JavaObjectArray<>)) {
+						var r = GetTypeSignature (type.GenericTypeArguments [0]);
+						return r.AddArrayRank (rank + 1);
+					}
+				}
+
+				var simpleRef = GetSimpleReference (type);
+				if (simpleRef != null)
+					return new JniTypeSignature (simpleRef, rank, false);
+
+				if (isGeneric) {
+					simpleRef = GetSimpleReference (genericDef);
+					if (simpleRef != null)
+						return new JniTypeSignature (simpleRef, rank, false);
+				}
+
+				return default;
```

Comment:
Why `return default` and not `return null`?

Reply:
I think because it is a `struct`? https://github.com/xamarin/java.interop/blob/5fa8c18b084fbeb16447e5070a12c32358f6245a/src/Java.Interop/Java.Interop/JniTypeSignature.cs#L12

I might also could do `new JniTypeSignature()` if that is more readable.

Reply:
Ah, right.  Duh.